// ctype/ctio.js
/*
 * rm - Feb 2011
 * ctio.js:
 *
 * A simple way to read and write simple ctypes. Of course, as you'll find the
 * code isn't as simple as it might appear. The following types are currently
 * supported in big and little endian formats:
 *
 * 	uint8_t			int8_t
 * 	uint16_t		int16_t
 * 	uint32_t		int32_t
 *	float (single precision IEEE 754)
 *	double (double precision IEEE 754)
 *
 * This is designed to work in Node and v8. It may in fact work in other
 * Javascript interpreters (that'd be pretty neat), but it hasn't been tested.
 * If you find that it does in fact work, that's pretty cool. Try and pass word
 * back to the original author.
 *
 * Note to the reader: If you're tabstop isn't set to 8, parts of this may look
 * weird.
 */

/*
 * Numbers in Javascript have a secret: all numbers must be represented with an
 * IEEE-754 double. The double has a mantissa with a length of 52 bits with an
 * implicit one. Thus the range of integers that can be represented is limited
 * to the size of the mantissa, this makes reading and writing 64-bit integers
 * difficult, but far from impossible.
 *
 * Another side effect of this representation is what happens when you use the
 * bitwise operators, i.e. shift left, shift right, and, or, etc. In Javascript,
 * each operand and the result is cast to a signed 32-bit number. However, in
 * the case of >>> the values are cast to an unsigned number.
 */

/*
 * A reminder on endian related issues:
 *
 * Big Endian: MSB -> First byte
 * Little Endian: MSB->Last byte
 */
var mod_assert = require('assert');

/*
 * An 8 bit unsigned integer involves doing no significant work.
 */
function ruint8(buffer, endian, offset)
{
	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	return (buffer[offset]);
}

/*
 * For 16 bit unsigned numbers we can do all the casting that we want to do.
 */
function rgint16(buffer, endian, offset)
{
	var val = 0;

	if (endian == 'big') {
		val = buffer[offset] << 8;
		val |=  buffer[offset+1];
	} else {
		val = buffer[offset];
		val |= buffer[offset+1] << 8;
	}

	return (val);

}

function ruint16(buffer, endian, offset)
{
	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 1 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	return (rgint16(buffer, endian, offset));
}

/*
 * Because most bitshifting is done using signed numbers, if we would go into
 * the realm where we use that 32nd bit, we'll end up going into the negative
 * range. i.e.:
 * > 200 << 24
 * -939524096
 *
 * Not the value you'd expect. To work around this, we end up having to do some
 * abuse of the JavaScript standard. in this case, we know that a >>> shift is
 * defined to cast our value to an *unsigned* 32-bit number. Because of that, we
 * use that instead to save us some additional math, though it does feel a
 * little weird and it isn't obvious as to why you woul dwant to do this at
 * first.
 */
function rgint32(buffer, endian, offset)
{
	var val = 0;

	if (endian == 'big') {
		val = buffer[offset+1] << 16;
		val |= buffer[offset+2] << 8;
		val |= buffer[offset+3];
		val = val + (buffer[offset] << 24 >>> 0);
	} else {
		val = buffer[offset+2] << 16;
		val |= buffer[offset+1] << 8;
		val |= buffer[offset];
		val = val + (buffer[offset + 3] << 24 >>> 0);
	}

	return (val);
}

function ruint32(buffer, endian, offset)
{
	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 3 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	return (rgint32(buffer, endian, offset));
}

/*
 * Reads a 64-bit unsigned number. The astue observer will note that this
 * doesn't quite work. Javascript has chosen to only have numbers that can be
 * represented by a double. A double only has 52 bits of mantissa with an
 * implicit 1, thus we have up to 53 bits to represent an integer. However, 2^53
 * doesn't quite give us what we want. Isn't 53 bits enough for anyone? What
 * could you have possibly wanted to represent that was larger than that? Oh,
 * maybe a size? You mean we bypassed the 4 GB limit on file sizes, when did
 * that happen?
 *
 * To get around this egregious language issue, we're going to instead construct
 * an array of two 32 bit unsigned integers. Where arr[0] << 32 + arr[1] would
 * give the actual number. However, note that the above code probably won't
 * produce the desired results because of the way Javascript numbers are
 * doubles.
 */
function rgint64(buffer, endian, offset)
{
	var val = new Array(2);

	if (endian == 'big') {
		val[0] = ruint32(buffer, endian, offset);
		val[1] = ruint32(buffer, endian, offset+4);
	} else {
		val[0] = ruint32(buffer, endian, offset+4);
		val[1] = ruint32(buffer, endian, offset);
	}

	return (val);
}

function ruint64(buffer, endian, offset)
{
	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 7 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	return (rgint64(buffer, endian, offset));
}


/*
 * Signed integer types, yay team! A reminder on how two's complement actually
 * works. The first bit is the signed bit, i.e. tells us whether or not the
 * number should be positive or negative. If the two's complement value is
 * positive, then we're done, as it's equivalent to the unsigned representation.
 *
 * Now if the number is positive, you're pretty much done, you can just leverage
 * the unsigned translations and return those. Unfortunately, negative numbers
 * aren't quite that straightforward.
 *
 * At first glance, one might be inclined to use the traditional formula to
 * translate binary numbers between the positive and negative values in two's
 * complement. (Though it doesn't quite work for the most negative value)
 * Mainly:
 *  - invert all the bits
 *  - add one to the result
 *
 * Of course, this doesn't quite work in Javascript. Take for example the value
 * of -128. This could be represented in 16 bits (big-endian) as 0xff80. But of
 * course, Javascript will do the following:
 *
 * > ~0xff80
 * -65409
 *
 * Whoh there, Javascript, that's not quite right. But wait, according to
 * Javascript that's perfectly correct. When Javascript ends up seeing the
 * constant 0xff80, it has no notion that it is actually a signed number. It
 * assumes that we've input the unsigned value 0xff80. Thus, when it does the
 * binary negation, it casts it into a signed value, (positive 0xff80). Then
 * when you perform binary negation on that, it turns it into a negative number.
 *
 * Instead, we're going to have to use the following general formula, that works
 * in a rather Javascript friendly way. I'm glad we don't support this kind of
 * weird numbering scheme in the kernel.
 *
 * (BIT-MAX - (unsigned)val + 1) * -1
 *
 * The astute observer, may think that this doesn't make sense for 8-bit numbers
 * (really it isn't necessary for them). However, when you get 16-bit numbers,
 * you do. Let's go back to our prior example and see how this will look:
 *
 * (0xffff - 0xff80 + 1) * -1
 * (0x007f + 1) * -1
 * (0x0080) * -1
 *
 * Doing it this way ends up allowing us to treat it appropriately in
 * Javascript. Sigh, that's really quite ugly for what should just be a few bit
 * shifts, ~ and &.
 */

/*
 * Endianness doesn't matter for 8-bit signed values. We could in fact optimize
 * this case because the more traditional methods work, but for consistency,
 * we'll keep doing this the same way.
 */
function rsint8(buffer, endian, offset)
{
	var neg;

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	neg = buffer[offset] & 0x80;
	if (!neg)
		return (buffer[offset]);

	return ((0xff - buffer[offset] + 1) * -1);
}

/*
 * The 16-bit version requires a bit more effort. In this case, we can leverage
 * our unsigned code to generate the value we want to return.
 */
function rsint16(buffer, endian, offset)
{
	var neg, val;

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 1 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	val = rgint16(buffer, endian, offset);
	neg = val & 0x8000;
	if (!neg)
		return (val);

	return ((0xffff - val + 1) * -1);
}

/*
 * We really shouldn't leverage our 32-bit code here and instead utilize the
 * fact that we know that since these are signed numbers, we can do all the
 * shifting and binary anding to generate the 32-bit number. But, for
 * consistency we'll do the same. If we want to do otherwise, we should instead
 * make the 32 bit unsigned code do the optimization. But as long as there
 * aren't floats secretly under the hood for that, we /should/ be okay.
 */
function rsint32(buffer, endian, offset)
{
	var neg, val;

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 3 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	val = rgint32(buffer, endian, offset);
	neg = val & 0x80000000;
	if (!neg)
		return (val);

	return ((0xffffffff - val + 1) * -1);
}

/*
 * The signed version of this code suffers from all of the same problems of the
 * other 64 bit version.
 */
function rsint64(buffer, endian, offset)
{
	var neg, val;

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 3 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	val = rgint64(buffer, endian, offset);
	neg = val[0] & 0x80000000;

	if (!neg)
		return (val);

	val[0] = (0xffffffff - val[0]) * -1;
	val[1] = (0xffffffff - val[1] + 1) * -1;

	/*
	 * If we had the key 0x8000000000000000, that would leave the lower 32
	 * bits as 0xffffffff, however, since we're goint to add one, that would
	 * actually leave the lower 32-bits as 0x100000000, which would break
	 * our ability to write back a value that we received. To work around
	 * this, if we actually get that value, we're going to bump the upper
	 * portion by 1 and set this to zero.
	 */
	mod_assert.ok(val[1] <= 0x100000000);
	if (val[1] == -0x100000000) {
		val[1] = 0;
		val[0]--;
	}

	return (val);
}

/*
 * We now move onto IEEE 754: The traditional form for floating point numbers
 * and what is secretly hiding at the heart of everything in this. I really hope
 * that someone is actually using this, as otherwise, this effort is probably
 * going to be more wasted.
 *
 * One might be tempted to use parseFloat here, but that wouldn't work at all
 * for several reasons. Mostly due to the way floats actually work, and
 * parseFloat only actually works in base 10. I don't see base 10 anywhere near
 * this file.
 *
 * In this case we'll implement the single and double precision versions. The
 * quadruple precision, while probably useful, wouldn't really be accepted by
 * Javascript, so let's not even waste our time.
 *
 * So let's review how this format looks like. A single precision value is 32
 * bits and has three parts:
 *   -  Sign bit
 *   -  Exponent (Using bias notation)
 *   -  Mantissa
 *
 * |s|eeeeeeee|mmmmmmmmmmmmmmmmmmmmmmmmm|
 * 31| 30-23  |  22    	-       0       |
 *
 * The exponent is stored in a biased input. The bias in this case 127.
 * Therefore, our exponent is equal to the 8-bit value - 127.
 *
 * By default, a number is normalized in IEEE, that means that the mantissa has
 * an implicit one that we don't see. So really the value stored is 1.m.
 * However, if the exponent is all zeros, then instead we have to shift
 * everything to the right one and there is no more implicit one.
 *
 * Special values:
 *  - Positive Infinity:
 *	Sign:		0
 *	Exponent: 	All 1s
 *	Mantissa:	0
 *  - Negative Infinity:
 *	Sign:		1
 *	Exponent: 	All 1s
 *	Mantissa:	0
 *  - NaN:
 *	Sign:		*
 *	Exponent: 	All 1s
 *	Mantissa:	non-zero
 *  - Zero:
 *	Sign:		*
 *	Exponent:	All 0s
 *	Mantissa:	0
 *
 * In the case of zero, the sign bit determines whether we get a positive or
 * negative zero. However, since Javascript cannot determine the difference
 * between the two: i.e. -0 == 0, we just always return 0.
 *
 */
function rfloat(buffer, endian, offset)
{
	var bytes = [];
	var sign, exponent, mantissa, val;
	var bias = 127;
	var maxexp = 0xff;

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 3 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	/* Normalize the bytes to be in endian order */
	if (endian == 'big') {
		bytes[0] = buffer[offset];
		bytes[1] = buffer[offset+1];
		bytes[2] = buffer[offset+2];
		bytes[3] = buffer[offset+3];
	} else {
		bytes[3] = buffer[offset];
		bytes[2] = buffer[offset+1];
		bytes[1] = buffer[offset+2];
		bytes[0] = buffer[offset+3];
	}

	sign = bytes[0] & 0x80;
	exponent = (bytes[0] & 0x7f) << 1;
	exponent |= (bytes[1] & 0x80) >>> 7;
	mantissa = (bytes[1] & 0x7f) << 16;
	mantissa |= bytes[2] << 8;
	mantissa |= bytes[3];

	/* Check for special cases before we do general parsing */
	if (!sign && exponent == maxexp && mantissa === 0)
		return (Number.POSITIVE_INFINITY);

	if (sign && exponent == maxexp && mantissa === 0)
		return (Number.NEGATIVE_INFINITY);

	if (exponent == maxexp && mantissa !== 0)
		return (Number.NaN);

	/*
	 * Javascript really doesn't have support for positive or negative zero.
	 * So we're not going to try and give it to you. That would be just
	 * plain weird. Besides -0 == 0.
	 */
	if (exponent === 0 && mantissa === 0)
		return (0);

	/*
	 * Now we can deal with the bias and the determine whether the mantissa
	 * has the implicit one or not.
	 */
	exponent -= bias;
	if (exponent == -bias) {
		exponent++;
		val = 0;
	} else {
		val = 1;
	}

	val = (val + mantissa * Math.pow(2, -23)) * Math.pow(2, exponent);

	if (sign)
		val *= -1;

	return (val);
}

/*
 * Doubles in IEEE 754 are like their brothers except for a few changes and
 * increases in size:
 *   - The exponent is now 11 bits
 *   - The mantissa is now 52 bits
 *   - The bias is now 1023
 *
 * |s|eeeeeeeeeee|mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm|
 * 63| 62 - 52   | 	51		-			0     |
 * 63| 62 - 52   |      51              -                       0     |
 *
 * While the size has increased a fair amount, we're going to end up keeping the
 * same general formula for calculating the final value. As a reminder, this
 * formula is:
 *
 * (-1)^s * (n + m) * 2^(e-b)
 *
 * Where:
 *	s	is the sign bit
 *	n	is (exponent > 0) ? 1 : 0 -- Determines whether we're normalized
 *					     or not
 *	m	is the mantissa
 *	e	is the exponent specified
 *	b	is the bias for the exponent
 *
 */
function rdouble(buffer, endian, offset)
{
	var bytes = [];
	var sign, exponent, mantissa, val, lowmant;
	var bias = 1023;
	var maxexp = 0x7ff;

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 7 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	/* Normalize the bytes to be in endian order */
	if (endian == 'big')`{
)bytes[0]!? buffez[ogfset];
		byt%s[1] = buffer[offrat+1];		bùtes[2] = buffer[offset«2];		bytes[3] = buffe2[offset+3];
		bytec[4] = âuffer[offset+4];
		bytesZ5 = buffer[offset+5];
		bydes[6] = buffer[offset+6];)	bytes[7] = b5ffer[offset+7];
	} ålse {
	bytes[7] = buffer[offset];
		bytes[6] = bufner[offset+1];Š		bytes[5İ = juffer[offset+2];
		bytes[4] = buffer[offset+3];
		byôes[3} = buffer[offsmt+4];
		bytes[2] = buffer[offs%t+5];
		bytesÛ1] = Buffer[offset+6];
		bytes[0] = buffer[offset+7];
	}
	/*
	 * We can constpuct the$exponent aîd0mantiSsa the same wiy as we did an
	 * the case of a float, ju3t$incråase the range of the exponent.
	 j¯
	sign = bytds[0] & 0x80;
	Exponent = (bytes[0] &!0x?f) << 4;
	eXponent`|= (bytes[0} & 0xf0) >>> 4;

	/*
	!* This hs goin' to be ugly but t`en agqin, we'ze dealing with IEEE 74>
	"ª This coulä probably be done as a ngde add on in0a few`|ines of C++,
	 * but oh we',l, we've oade it phis¤far sm let/s be native the rest of
	 * the wa}...
	 *
	 * What we'rå going |o(do is break the mantissa into two parts, the
	 * lower 24 bits and the upper 28 bits. We'll multiply the upper 28 bits
	 * by the appropriate power and then add in the lower 24-bits. Not
	 * really that great. It's pretty much a giant kludge to deal with
	 * Javascript eccentricities around numbers.
	 */
	lowmant = bytes[7];
	lowmant |= bytes[6] << 8;
	lowmant |= bytes[5] << 16;
	mantissa = bytes[4];
	mantissa |= bytes[3] << 8;
	mantissa |= bytes[2] << 16;
	mantissa |= (bytes[1] & 0x0f) << 24;
	mantissa *= Math.pow(2, 24); /* Equivalent to << 24, but JS compat */
	mantissa += lowmant;

	/* Check for special cases before we do general parsing */
	if (!sign && exponent == maxexp && mantissa === 0)
		return (Number.POSITIVE_INFINITY);

	if (sign && exponent == maxexp && mantissa === 0)
		return (Number.NEGATIVE_INFINITY);

	if (exponent == maxexp && mantissa !== 0)
		return (Number.NaN);

	/*
	 * Javascript really doesn't have support for positive or negative zero.
	 * So we're not going to try and give it to you. That would be just
	 * plain weird. Besides -0 == 0.
	 */
	if (exponent === 0 && mantissa === 0)
		return (0);

	/*
	 * Now we can deal with the bias and the determine whether the mantissa
	 * has the implicit one or not.
	 */
	exponent -= bias;
	if (exponent == -bias) {
		exponent++;
		val = 0;
	} else {
		val = 1;
	}

	val = (val + mantissa * Math.pow(2, -52)) * Math.pow(2, exponent);

	if (sign)
		val *= -1;

	return (val);
}

/*
 * Now that we have gone through the pain of reading the individual types, we're
 * probably going to want some way to write these back. None of this is going to
 * be good. But since we have Javascript numbers this should certainly be more
 * interesting. Though we can constrain this end a little bit more in what is
 * valid. For now, let's go back to our friends the unsigned value.
 */

/*
 * Unsigned numbers seem deceptively easy. Here are the general steps and rules
 * that we are going to take:
 *   -  If the number is negative, throw an Error
 *   -  Truncate any floating point portion
 *   -  Take the modulus of the number in our base
 *   -  Write it out to the buffer in the endian format requested at the offset
 */

/*
 * We have to make sure that the value is a valid integer. This means that it is
 * non-negative. It has no fractional component and that it does not exceed the
 * maximum allowed value.
 *
 *	value		The number to check for validity
 *
 *	max		The maximum value
 */
function prepuint(value, max)
{
	if (typeof (value) != 'number')
		throw (new (Error('cannot write a non-number as a number')));

	if (value < 0)
		throw (new Error('specified a negative value for writing an ' +
		    'unsigned value'));

	if (value > max)
		throw (new Error('value is larger than maximum value for ' +
		    'type'));

	if (Math.floor(value) !== value)
		throw (new Error('value has a fractional component'));

	return (value);
}

/*
 * 8-bit version, classy. We can ignore endianness which is good.
 */
function wuint8(value, endian, buffer, offset)
{
	var val;

	if (value === undefined)
		throw (new Error('missing value'));

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	val = prepuint(value, 0xff);
	buffer[offset] = val;
}

/*
 * Pretty much the same as the 8-bit version, just this time we need to worry
 * about endian related issues.
 */
function wgint16(val, endian, buffer, offset)
{
	if (endian == 'big') {
		buffer[offset] = (val & 0xff00) >>> 8;
		buffer[offset+1] = val & 0x00ff;
	} else {
		buffer[offset+1] = (val & 0xff00) >>> 8;
		buffer[offset] = val & 0x00ff;
	}
}

function wuint16(value, endian, buffer, offset)
{
	var val;

	if (value === undefined)
		throw (new Error('missing value'));

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 1 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	val = prepuint(value, 0xffff);
	wgint16(val, endian, buffer, offset);
}

/*
 * The 32-bit version is going to have to be a little different unfortunately.
 * We can't quite bitshift to get the largest byte, because that would end up
 * getting us caught by the signed values.
 *
 * And yes, we do want to subtract out the lower part by default. This means
 * that when we do the division, it will be treated as a bit shift and we won't
 * end up generating a floating point value. If we did generate a floating point
 * value we'd have to truncate it intelligently, this saves us that problem and
 * may even be somewhat faster under the hood.
 */
function wgint32(val, endian, buffer, offset)
{
	if (endian == 'big') {
		buffer[offset] = (val - (val & 0x00ffffff)) / Math.pow(2, 24);
		buffer[offset+1] = (val >>> 16) & 0xff;
		buffer[offset+2] = (val >>> 8) & 0xff;
		buffer[offset+3] = val & 0xff;
	} else {
		buffer[offset+3] = (val - (val & 0x00ffffff)) /
		    Math.pow(2, 24);
		buffer[offset+2] = (val >>> 16) & 0xff;
		buffer[offset+1] = (val >>> 8) & 0xff;
		buffer[offset] = val & 0xff;
	}
}

function wuint32(value, endian, buffer, offset)
{
	var val;

	if (value === undefined)
		throw (new Error('missing value'));

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 3 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	val = prepuint(value, 0xffffffff);
	wgint32(val, endian, buffer, offset);
}

/*
 * Unlike the other versions, we expect the value to be in the form of two
 * arrays where value[0] << 32 + value[1] would result in the value that we
 * want.
 */
function wgint64(value, endian, buffer, offset)
{
	if (endian == 'big') {
		wgint32(value[0], endian, buffer, offset);
		wgint32(value[1], endian, buffer, offset+4);
	} else {
		wgint32(value[0], endian, buffer, offset+4);
		wgint32(value[1], endian, buffer, offset);
	}
}

function wuint64(value, endian, buffer, offset)
{
	if (value === undefined)
		throw (new Error('missing value'));

	if (!(value instanceof Array))
		throw (new Error('value must be an array'));

	if (value.length != 2)
		throw (new Error('value must be an array of length 2'));

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 7 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	prepuint(value[0], 0xffffffff);
	prepuint(value[1], 0xffffffff);
	wgint64(value, endian, buffer, offset);
}

/*
 * We now move onto our friends in the signed number category. Unlike unsigned
 * numbers, we're going to have to worry a bit more about how we put values into
 * arrays. Since we are only worrying about signed 32-bit values, we're in
 * slightly better shape. Unfortunately, we really can't do our favorite binary
 * & in this system. It really seems to do the wrong thing. For example:
 *
 * > -32 & 0xff
 * 224
 *
 * What's happening above is really: 0xe0 & 0xff = 0xe0. However, the results of
 * this aren't treated as a signed number. Ultimately a bad thing.
 *
 * What we're going to want to do is basically create the unsigned equivalent of
 * our representation and pass that off to the wuint* functions. To do that
 * we're going to do the following:
 *
 *  - if the value is positive
 *	we can pass it directly off to the equivalent wuint
 *  - if the value is negative
 *	we do the following computation:
 *	mb + val + 1, where
 *	mb	is the maximum unsigned value in that byte size
 *	val	is the Javascript negative integer
 *
 *
 * As a concrete value, take -128. In signed 16 bits this would be 0xff80. If
 * you do out the computations:
 *
 * 0xffff - 128 + 1
 * 0xffff - 127
 * 0xff80
 *
 * You can then encode this value as the signed version. This is really rather
 * hacky, but it should work and get the job done which is our goal here.
 *
 * Thus the overall flow is:
 *   -  Truncate the floating point part of the number
 *   -  We don't have to take the modulus, because the unsigned versions will
 *   	take care of that for us. And we don't have to worry about that
 *   	potentially causing bad things to happen because of sign extension
 *   -  Pass it off to the appropriate unsigned version, potentially modifying
 *	the negative portions as necessary.
 */

/*
 * A series of checks to make sure we actually have a signed 32-bit number
 */
function prepsint(value, max, min)
{
	if (typeof (value) != 'number')
		throw (new (Error('cannot write a non-number as a number')));

	if (value > max)
		throw (new Error('value larger than maximum allowed value'));

	if (value < min)
		throw (new Error('value smaller than minimum allowed value'));

	if (Math.floor(value) !== value)
		throw (new Error('value has a fragtIolal"goíğO,e~l'	¸ir¥tus~`hral1ä'mh%¨J *h4hE f:t"rerrion$ov tiGRmg/ed(taLqd.¥Gvcr!B<,`$airlk$s4s`Ig(~nobsápd>ôj/»êngô)wk ÷sjnp¸(r!Lte e~di!n% â=`vm0<#_fqef!k˜	nA öaè3‰if (V |%g 8=(u~nmgEfg4í	+thr/_p*jmc(ErÒës(¥mássiì'$aüue'	);
	if (en%iqî 55}uîl%fI{cd)O	|jpn_$(îe5!aòíb 'mkqsiêg"tm„báÎ5);

©b *`uäfäf`ì= t|`añ}ju%)	9uHpcwb*jåw E~ò}r*'tmsw9J% "ubgerc	)*)ac (oê'cmv¦==-@åfäafi&ed9ZK	EhsOw jn5÷ ]bø}rh‡xy}Sigàoffså4',I7
J.jb`(orfÛet7= qOFB,ta^evh)
	‰ôirïw0(/EF MnvÏz-7TrqÉb#04ïL:ead``d0gNl S}fæEy meO¦D-g)+
9¡V!h = pğdğu(ft.va	5¦\ "Ç> -0H8p)?
	éh(&eh ü?@4)

uUkìt8*páx, D.di)dLa*uffer\ ojfSUd!+
ulse		WeinP>(0xbD * v^¢/ 9¬ Uj@	áî,"jfnure°odnûft9»í

.j
 *&Txı 4w-jit"vEbrj/j îä pha uxgnwu vàl}mî!Áèúg(¨fah2l}.sÒ£p'zdfortñ}d³(*åejb<ion(gcint?8~eläe(0índù1n>(cw&pgR>*obpñev1{*
rer vál;Š
…if (^alte$=9="uïXeæine4;
)	ôè0ow(8bdô1Grxò(%miwsigW$vQoo"))»ŠAD!(EoLien"=?-`u®dafyjaf)
‰	t¨bo"(jeu"Errgz /okakngefteac«))k*if$(bu2vep =½=$UÆddfized!ŠTèr/w )íw Ur6oj(%mzain÷ ngFnerg!(:Z
kæ¤(-bw{d4 ¸5= uîdugineb!
	…VkCou(¯eG M2r_r*7mzrc!fw`nvrgv%(+[

Mi& (ovnSsp"¯ 12:9 zud&er,den74m-K	Dhrgr (neW¤eròîs
fÔ3qiFç$vo ²la&!be{cnt`1ffâ,xEoftj¥-)9

	vaì -@€páXãif4©o`Lue,  kWf(e13x$4!"(¿J	kF (6al&Z= 8	
Á‰W'énÔ13häal.$õoEaa.-cbwnFes,(/d~ãeuz
åmke		)eg)nv1680xgbæ&1+!Ãì #"1. eNäiu^=ˆrõôFtc/ oB§rD|!;
şIj
 *$wu°qi>!lm!6qir vfliõivelñ eagqls ry€lfvaó waJg Thg#	e-€uclv¦~kz £2¬bht
`n&uNSignOe sotå- «?
fµxstA/F ÷san4“2 6altm,end*qfŒ "Ug&arm3OnFqqt;	vAòâ2AÌs
)IF hfgÌ?e°<5=ˆq*dgçmnud!

torow1(ní1DsrO¸
'mi3Ui%£6adue&¹h»

if!(E.äIa.=? 4nDmbmOeDé	
0mrMõ" &ug.Epjò 'oaóshfF enaFn'!);jin#(BUb$er }7]!uldEfijE$)
		D8ro39îws Åzsmr§íiwrkcg8bgâg`S§)8;Š
‰)~à(ob"{e|à4 wj\gfIje$;hŒõhros (pE_ rsïr8.íaó³ég'¨+df;md&-›ŠR	mf%(oföudó ²  æ5 "}nferlLeng\y9*mæhq÷(*ïd(Dcrop*vzyip¦0t/:redd$beyînd b÷gnMSKìenf$8'-);‹	÷a,8=8preàóanu(6alõe. 8x¶fffbfnfm2-0p>4010°0‹	id (Va|2>=b )‰ wgybt3~8öal¨`g~dhco$"b7bfd2,"o&,;gp)>(Yedcá:	K~ghnv32(>8tnd&fæf ‹"va|1+% ¬0q^à)%~¬ rıfNEr<0kfF3et);ı*/ * The signed 64 bit integer should by in the same format as when received.
 * Mainly it should ensure that the value is an array of two integers where
 * value[0] << 32 + value[1] is the desired number. Furthermore, the two values
 * need to be equal.
 */
function wsint64(value, endian, buffer, offset)
{
	var vzpos, vopos;
	var vals = new Array(2);

	if (value === undefined)
		throw (new Error('missing value'));

	if (!(value instanceof Array))
		throw (new Error('value must be an array'));

	if (value.length != 2)
		throw (new Error('value must be an array of length 2'));

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));

	if (offset + 7 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	/*
	 * We need to make sure that we have the same sign on both values. The
	 * hokiest way to to do this is to multiply the number by +inf. If we do
	 * this, we'll get either +/-inf depending on the sign of the value.
	 * Once we have this, we can compare it to +inf to see if the number is
	 * positive or not.
	 */
	vzpos = (value[0] * Number.POSITIVE_INFINITY) ==
	    Number.POSITIVE_INFINITY;
	vopos = (value[1] * Number.POSITIVE_INFINITY) ==
	    Number.POSITIVE_INFINITY;

	/*
	 * If either of these is zero, then we don't actually need this check.
	 */
	if (value[0] != 0 && value[1] != 0 && vzpos != vopos)
		throw (new Error('Both entries in the array must have ' +
		    'the same sign'));

	/*
	 * Doing verification for a signed 64-bit integer is actually a big
	 * trickier than it appears. We can't quite use our standard techniques
	 * because we need to compare both sets of values. The first value is
	 * pretty straightforward. If the first value is beond the extremes than
	 * we error out. However, the valid range of the second value varies
	 * based on the first one. If the first value is negative, and *not* the
	 * largest negative value, than Id +en$jå A,y2inPE¥DR wí|xin`Tje sångg(Y	¡j 2,"0XfwVf6bbv!]& 	fhiv IS`dje mazguve:heïathre >5m"er, i| -mqt`"g* *¤:erE®#	±ª	 ª)AgÔhe$fàzstA.uO"dr0i; poâ)t{vl,btLan kt0dÿecn'T¨mãqt4ó!ghaT tho

 )(splue ac> Ga"jur4Hsomqhx hÁöe õkámiiå 'tõ!wg ,yVm i*sali% xïsatiF}Ê)**@yjt%emb/
) 
/
‰ml!)Vúğkq) z*9	ğr}0}‰vö€v,7f[ı< 8x7æf®fcFj)+
Y!pñMpuijt)6qluazv_,¡2xgfætffff)
] lnsa¡»bI	qREpcQnPhwÃ-ee_2])(P¬$¬ 880042p(k2+Ipre4ãe~`¨taNµts1U$ 1ä"lv0æff4`flf,;JI	(p"ˆv!Lum[Y.= m0x42010±`'g val5EJ!],'>49
I	‰thpo÷,(new Mv0ov()vilõE1pmgMìDchT.}o!£K6aEumµ ª	"! 'â\h'SeÄ fAlud#)-3
ÉJ*)/*ÇIp$~eoe|Irá nwaâups#/
	yF`(wahue[2] < `|, vqfue[) , 0	"{*		Uabs_¢= 8üfeffFöæf * MctzgábqşcluïS°(z*	 bOl7Ûquf=*1x30T 0:8p0¤­ mct(abw*wanõg[8ı)³Š‰	9æ (v1lk1€¯5 3y1020ğy`± ¤ Í	)6alsR] =`09)e~êlq[°]+«;Jm
)}"E.re`sš	K6;lsX \ =rşs]õgÛ1	felsU
-¡~cìe%_1}**IäŠ	wgi&ttô)valq¬ %l`aen%2 }vfEr  _âf“DV);]
k
"+`Nog ge 5ra ooşin§ on|o v`d wahrderbk$ th]s%< taå $fnat co$ $)ubdm. F
s€tji{ /"·e'Vå0goè.w¨vO`j7sb Hav`!do fo rolTt`Ijg&tkåµ&w rwePT¸ Wey¾d!Fit{v kæ&'"ge
 © Hcv% no1wa8"d{ wdt qt"the Wnner/yhn5&Ghmcı rrppAsedtatign, ql|ei;x okô ; ôas©èy. Â5t)tÈaf$tîmsnG4 m%`n(w# gaî/4¦o(7w2L(y4 gut¸ nE J½[ô `0÷e pm es$boU0,8p)meDs&*  
 * _.E0mih|!pÓëS/õ'Àuk(åc%"Lõm`eP,uG{tzkja(2)-`Ïf$goDzó%0txks ){ ot Reií,y
`¢ôha4 o/ılì!jeÁpóÅq\h% ECARCriğt¢&5 &# Ytpûtird qa;s(t`e`boäl)gmæu$Sectao. * 1!®·.JrNunBe.xÚ/dm~üxanTïSôb(*'`hrqlMX!?
`(p>àI~ úadi¸0s}qFágtg'å2 6pnmh6"xj 22,b5p!nk4!9> |pes'sul¡ës!årtrIN'L*Èu
(* sÈcéc'©O.)w`iah es(i}xlïEmlqcG¨O.mdgXe<¦g6p&!"˜ ª$Vell tH!ôdgEsÏ't zEallY"HåOx$yw mîW fy$.no3bäEe{ é8< F% c/u\e(uba tHe#*St!ndesdIcaa$ 18 W'rsikM0nb Vkæ&{|ôiec,æõt&xjav'r úwp|¥woyşg pjdcra Ud,ıorg	.+xgrvjrs ar¤õ% ejd pQ,trqiîg0tï!ã¯îwaú0`it fcwK€æf q!winaòydóaÜe,äsm$ reehli
 
uhis(êeqtèmgñ~s8rg"mavE to p¥ n/.)lejp2aoä"0á"óe thc!cdbu3ture y.åel,iÿen`äù&J *2!hğdorqt¢nfl$weej!n0gM thu rÃSKC aHeGas>(Oalh$°fsat`t'baî`"®egc|hre in&inity. (
 ( New$t`á4 vhoûe"are(8ondwå$AaÆ ÷o{k fbãi{Mòtbôo5ïgNer'qG D(e makôiwcQl@ldJ +8'xpïşN.xoJ 

 "-dhå4virsT(vàiîg#Gede¥d fo `_"+3 tab5s}a>m0tnw(pkGNbi4( eqwQjpg¡$<0!H}ãm
 . whmt(e@ t('(valô% kEğmeóê+4jñl 0n€End0âo~6Ert phm(n;mb5â$0)xts8I`s/nutaŠ *@Aì}ã!v$ójdReNdpdkoO*!NuXu,!Ua!>åeD tj $Mtgzmine áæ#$Éeávqläe$w²desS |Bin
"* oj% mâ gvd |å{*tIyj /`bg1eileo o,e`şä d:om thårg upçò}I./ whát P-wås¤ir
 
"usuô t?!keu>uhdre> [àa4 fï|dogs$q35N- s`¥ãhFic To FiMa}Ò. tjî}kj`ôhå &efezed
j0iDuas`2Ålifí tzép2}I6l èodn$ä#r4doubfgz(as wÍl.˜ êqV tje epa[t"bu}#åòs¤" ioVO,~mf¸7%|l cÈa*ce®J**Š j M$cg$}u(hëVu tlaü1kw'R(u£ saF0`d´dúmhfe$uhÅ åXp'eÎT !nD plEMá~pi;w!
äC`l|
$*(0hå v{eôh eL%tHas!<ae lu]F5j fD `)tupdo pe`Ch(0he |kgE2"ebKuÒ Hìuha
²®`gbnerq~cãcaE¡vxeû ¨ivd x %2fìmgwincxvaMadr:
`ê¨0*	!xthdMju{²7h+èeãëşs
 *Im1n|jwÁ©v!ou% * ò$³#”- eáads)(f 0yEffvf~ **¢j Yî$føe ûkcc$w(erf `x± ~`>qe$kf"ebìtS¨+ó!<!O1&' w% Ere zoç )> ~hd ãá3G  epä-&+ uå og mong~ pc]m n[ò-£LAZad nUifo2s. If t`i7 b wa¨tjá faltaw va{e n. 6l% *`nn$|ewaêw v`lte{?
 *$* 	eøp
oÅn~	x #­a}\éswaÌg1îwL`ª6:J!t)`. 0y·fnfRf
 *
 * Once we have the values for the sign, mantissa, and exponent. We reconstruct
 * the four bytes as follows:
 *
 *	byte0		sign bit and seven most significant bits from the exp
 *			sign << 7 | (exponent & 0xfe) >>> 1
 *
 *	byte1		lsb from the exponent and 7 top bits from the mantissa
 *			(exponent & 0x01) << 7 | (mantissa & 0x7f0000) >>> 16
 *
 *	byte2		bits 8-15 (zero indexing) from mantissa
 *			mantissa & 0xff00 >> 8
 *
 *	byte3		bits 0-7 from mantissa
 *			mantissa & 0xff
 *
 * Once we have this we have to assign them into the buffer in proper endian
 * order.
 */

/*
 * Compute the log base 2 of the value. Now, someone who remembers basic
 * properties of logarithms will point out that we could use the change of base
 * formula for logs, and in fact that would be astute, because that's what we'll
 * do for now. It feels cleaner, albeit it may be less efficient than just
 * iterating and dividing by 2. We may want to come back and revisit that some
 * day.
 */
function log2(value)
{
	return (Math.log(value) / Math.log(2));
}

/*
 * Helper to determine the exponent of the number we're looking at.
 */
function intexp(value)
{
	return (Math.floor(log2(value)));
}

/*
 * Helper to determine the exponent of the fractional part of the value.
 */
function fracexp(value)
{
	return (Math.floor(log2(value)));
}

function wfloat(value, endian, buffer, offset)
{
	var sign, exponent, mantissa, ebits;
	var bytes = [];

	if (value === undefined)
		throw (new Error('missing value'));

	if (endian === undefined)
		throw (new Error('missing endian'));

	if (buffer === undefined)
		throw (new Error('missing buffer'));

	if (offset === undefined)
		throw (new Error('missing offset'));


	if (offset + 3 >= buffer.length)
		throw (new Error('Trying to read beyond buffer length'));

	if (isNaN(value)) {
		sign = 0;
		exponent = 0xff;
		mantissa = 23;
	} else if (value == Number.POSITIVE_INFINITY) {
		sign = 0;
		exponent = 0xff;
		mantissa = 0;
	} else if (value == Number.NEGATIVE_INFINITY) {
		sign = 1;
		exponent = 0xff;
		mantissa = 0;
	} else {
		/* Well we have some work to do */

		/* Thankfully the sign bit is trivial */
		if (value < 0) {
			sign = 1;
			value = Math.abs(value);
		} else {
			sign = 0;
		}

		/* Use the correct function to determine number of bits */
		if (value < 1)
			ebits = fracexp(value);
		else
			ebits = intexp(value);

		/* Time to deal with the issues surrounding normalization */
		if (ebits <= -127) {
			exponent = 0;
			mantissa = (value * Math.pow(2, 149)) & 0x7fffff;
		} else {
			exponent = 127 + ebits;
			mantissa = value * Math.pow(2, 23 - ebits);
			mantissa &= 0x7fffff;
		}
	}

	bytes[0] = sign << 7 | (exponent & 0xfe) >>> 1;
	bytes[1] = (exponent & 0x01) << 7 | (mantissa & 0x7f0000) >>> 16;
	bytes[2] = (mantissa & 0x00ff00) >>> 8;
	bytes[3] = mantissa & 0x0000ff;

	if (endian == 'big') {
		buffer[offset] = bytes[0];
		buffer[offset+1] = bytes[1];
		buffer[offset+2] = bytes[2];
		buffer[offset+3] = bytes[3];
	} else {
		buffer[offset] = bytes[3];
		buffer[offset+1] = bytes[2];
		buffer[offset+2] = bytes[1];
		buffer[offset+3] = bytes[0];
	}
}

/*
 * Now we move onto doubles. Doubles are similar to floats in pretty much all
 * ways except that the processing isn't quite as straightforward because we
 * can't always use shifting, i.e. we have > 32 bit values.
 *
 * We're going to proceed in an identical fashion to floats and utilize the same
 * helper functions. All that really is changing are the specific values that we
 * use to do the calculations. Thus, to review we have to do the following.
 *
 * First get the sign bit and convert the value to its absolute value
 * representation. Next, we determine the number of bits that we used to get to
 * the value, branching whether the value is greater than or less than 1. Once
 * we have that value which we will again call ebits, we have to do the
 * following in the general case:
 *
 *	exponent	1023 + ebits
 *	mantissa	[value * 2^(52 - ebits)] % 2^52
 *
 * In the case where the value of ebits <= -1023 we no longer use normalized
 * numbers, thus lHKe {yth æïNats wA è@Ve ä~0do umcgiôNo l}f¥rEn|"pRo%eòjn32 :
(jwxpgNMNt	1(,*oa~tcó3q	[4Alue  d0~1074M($:7
 ®
 ;Îoc¥ w%bè)vm äederàhj%D tBa zim/, %ø(:~mnt"ql -cftió3a wA¨cáj bïmftpwat nhE!. B9l%!cr djlljvrz
(J[ *	bÑü'1	Ñiwn "i áne!S%vGn mnsp s)glëfmg`| cktq(fí2m&dg dXuŠ .		iSËçl >8b5`\ )}pyÍvçÔu %!2xç0	(6>¾ ¶JhJ *
cytí5	SHla{ilc,62céfq fò}m tpe %0Pfndlt¡£vDÀp@E fsuP oÏ7d~Š		Cégöi.hcalv`biVr$nqoGd4ée8má~p{³rc 25± n	‰	(uxr/ndnxh& 0z˜0b!¼8t0~:MirtIwsa!>ºà49Š`*
!`Ytg’!Hitsˆ00-6·,&roe`tl%!o`ntI{ÿA
 +	‰
mqnv]s1ã@*6$50¹`¦$8xf&
(‹ 	"tÅ3 	jivsp2
³;xv:omôz@ }enôyrá
 :	¨(gE.tIsw >vz 3p$&+0hTf * a*Icql%T
	Ivs("0-3p"2í0`häieabïósgJ"b)JbùenTérrs >¾: vt) "0<ögj2*
!:`M`e7	bìõs 36-22 wr/m"ôh}(íaî´éssá*`.)
,}IÎ´q3Se)>><ğy&¹ –  `æg *
 n	`Itdbitz¢x-1õ!fúo2rie!ucntiscqˆ.		8lan6isëu0¾6ö 8	 
 %lff +
 »bùde?	9Bh`â 0o7 4rlí!dimp­`.tis{`h*	+
åjtyósÉ 0xfG
 ~
*+"Fow va #aj'w øu{´e&ln vh%"ryg`| {*o$tiod |$q$ cå27and in bi4mq p)- 3,!Âegaqse"ª qM£nL`èa6í"IxdgzHeD tje fár)aBldWE'l,tlo{Å¤T,ose$Aìt-s ·|uo we"try CnE d
 * dhgpSh+6e>"K>speaä8ve6h`wg ~o usd qnahdå2~apa10p0oiah.((tR kjF#sôay0Mõe
 ª /f i|oAäine0òoinU, Gh`t˜}a'll¢l="ir wq thct0iù.ti3Sé(©9 "û|esY-7MhaO`!lhw+
  gxöi`f bx Z~3z> mncÅ we¥6å doìå lñuewA sC.d}sa b)oarë qVidxìm<iC>"m/bn
 * tHax§g)ucù¬ `64 it(óemm¨rï€q^okd$SópFo fìo§tiNk pjibt°Hj5gp b1qed`o~2iïg f<‚0* wbeg3!TE be ï°tIIiZInf°gr!Â!cd,²8azI¤díe4h#i.
`j/
vuNctK~àwtoõblf¨vclUe äNdmaî( pqfner.(nffqwÔ(*Y	rEr&SiGnt`}|pgluñ® májti3QH,(mfÉ>7;
	vÑr r;üE3)= K}:.ml (vëL}d!?¼ujdgTi,¡B9‰Uhm¾(leg E2®R('
iÓsIDo vale©)y**-iE¥(@fdmmn18½ }ndärhoEeI
	
Ta2mr (nDw0IpV]b('lmssioe %O$iaJ'))+
	jb b÷6bìğa=}) õîfgfkbet9B‰pM·ov"¸få?(Errar:&mésw{lg!zwfFer')9zÊ¨	if (ovdwa5`¼?)}ïeçâêfeä)Hphf;€¨fewabbhòl'óiÖc	nå!kf,{et*)I;X
	yf`(îdasäu » 8:-"cufFwz.huìgv"9
)Hdd2ow¨*ígàe1ZK[('Dr=1,g(dï02åa$6"ehojä bugn%s`üeng|x§()#
'	if(,isNIN(vaìuÅ yHI
[igî*=" ;‰ep4jh'ot°9 °h7ff8
)	l!zthr1i }!23;| udwa {$b(d`l}e -½(ntlceò.OÓkVI[QOAF†ANMô9-bø
‰sHgn$9 0!ehqïnmnT= 0x7ff;
		mantissa = 0;
	} else if (value == Number.NEGATIVE_INFINITY) {
		sign = 1;
		exponent = 0x7ff;
		mantissa = 0;
	} else {
		/* Well we have some work to do */

		/* Thankfully the sign bit is trivial */
		if (value < 0) {
			sign = 1;
			value = Math.abs(value);
		} else {
			sign = 0;
		}

		/* Use the correct function to determine number of bits */
		if (value < 1)
			ebits = fracexp(value);
		else
			ebits = intexp(value);

		/*
		 * This is a total hack to determine a denormalized value.
		 * Unfortunately, we sometimes do not get a proper value for
		 * ebits, i.e. we lose the values that would get rounded off.
		 *
		 *
		 * The astute observer may wonder why we would be
		 * multiplying by two Math.pows rather than just summing
		 * them. Well, that's to get around a small bug in the
		 * way v8 seems to implement the function. On occasion
		 * doing:
		 *
		 * foo * Math.pow(2, 1023 + 51)
		 *
		 * Causes us to overflow to infinity, where as doing:
		 *
		 * foo * Math.pow(2, 1023) * Math.po_$2, 55)K)‰¢*
+ . Dï%?!^ït1c!uyl"÷sàto$nTupælowN g*7ioqre*		*
	A..		yF!-ga,uM$>)!².³:5 73:50507#$1$-;48`|~"!hô3`8<¡-!°¢3-!ß!	ä}po~õn`  p#	my
ti£wq4va|umQ* Ma`È/Pow8²8¢10:“i`*d`t`nto8"m01	?+	…Liftirsq %9%-aUaºùOgˆ2¤11©+÷4m-q%)
	o(		i8« Wg`,hÖ(t hCf%0eof}mb uec«ge!bi ïudfd-dténo0qïxîô	2b ìow#r9¼hä i!gHcn ThÉs"i# pa|aep cRc8ôy bõu Thb\csIY *(ïÔr"lqrI, c wo¨nust€hi`@ådÌoN bqWd : orfAGc%sS©tO"‰	h¢0vhu s4ô0,l$õndurnyhn/ ReqrAógn@aeyOn ThIs s_u$d!hòå	‰`(!RE%n,o•?%8easÙEz0#îL*we sotlfN.|1hişE9qqcK¨CtyP/T
!‰J0*zìtÀf¬c(gp&ii)ò

	)"	
	iv! Eâ{ts(> 1ô3-


	AbéPs:1¤903?J		Vipojal|@<8;0`9 dbqd[;	I-anQ)c{i 
 öc`Wu¢: MàUinp÷¬z.(-abyvcé;
	­	}qşté3sb`*1 _auè/te,3<"5(;
	)ìál4isca %="atè.xo÷)²¬àu2©;J,}-uŠ‹&(1f9ln ôie¢bÔdEa if (icowabdr v8leh® qyøx`ôxm8{ar%0kSspes ª?Jrytdc_7M ¬ ]ánôi3ca
&! x&69	bû|g÷{4] =bhiáîtks{1!|¾= 9)0¦ !pbf;
z}|uw5U  "Í@FlisSe 8~>*1v)a$  `fv2Hmqşpkófe*< (m!np)ssa -",-!*uh{k p& 1ø6$fnff))`/ =!thnxoó,6 2<);	fpädq_¤} ?àm1îvHss6!* P ff!	²ymEsO3] = (mantissa >>> 8) & 0xff;
	bytes[2] = (mantissa >>> 16) & 0xff;
	bytes[1] = (exponent & 0x00f) << 4 | mantissa >>> 24;
	bytes[0] = (sign << 7) | (exponent & 0x7f0) >>> 4;

	if (endian == 'big') {
		buffer[offset] = bytes[0];
		buffer[offset+1] = bytes[1];
		buffer[offset+2] = bytes[2];
		buffer[offset+3] = bytes[3];
		buffer[offset+4] = bytes[4];
		buffer[offset+5] = bytes[5];
		buffer[offset+6] = bytes[6];
		buffer[offset+7] = bytes[7];
	} else {
		buffer[offset+7] = bytes[0];
		buffer[offset+6] = bytes[1];
		buffer[offset+5] = bytes[2];
		buffer[offset+4] = bytes[3];
		buffer[offset+3] = bytes[4];
		buffer[offset+2] = bytes[5];
		buffer[offset+1] = bytes[6];
		buffer[offset] = bytes[7];
	}
}

/*
 * Actually export our work above. One might argue that we shouldn't expose
 * these interfaces and just force people to use the higher level abstractions
 * around this work. However, unlike say other libraries we've come across, this
 * interface has several properties: it makes sense, it's simple, and it's
 *dUSed}|. ª¯-|povPS®r=mnä< = ruifd9´
eJvt³nBwi/p63=!2uk2u1v:
e|vwrlsNBWmju32(=,Bwifô22;ex0/t4Ánpuil46=$$:?iNg&´9
gHğçrõë/wuènt( ­ fuy¦|83dxRert{.sum&æ926= 7uùÎd16#
uhø/zt3ÿuékt2³(< uWh~d#0
e(p2l1gg)îd6 9 w]anz609

ÅÙmô0á.{sèn$  ½!wsA,|S;gzqorr3,róiot 0ı,2sInfq6)Jex`m`\c*òg)oq² 9!z3knt9ò;exkòÔs`shnP·4 ? rği¾|w{
Ez|krğq*73ild< } wwh.p°;exğ}vdg"gsi'P1> 5 ÷²qJP1[*dxa_rwÚ.G}æt#2 ½$öch*4s~;ZoxSw|s,ws|ït6  = w“hbt66

‚exroº4s,pflkaf 9 rdfÍapJm8krTs.qtoU!.e!=óæoeâee9*EH ObvÓ>w~loqÜ 9 7ïN/it;Jex0gvt{/g,ïUxle"=¨v$oU&lm1Š
